Финальный проект 4 спринта.

atestmirror opened the pull request:
Товарищ ревьюер, доброе утро!
Я совсем запуталась с заданием для проверки формы заказа самоката.
Как я поняла, некоторым элементам можно передать значение в форме ввода (например, станция метро или дата).
Но при запуске тестов оказалось, что это не так работает:)
```
public void setUndergroundStationField(String station) {
        driver.findElement(undergroundStation).sendKeys(station);
    }
```
На стаковерфлоу нашла метод, который присуждает значение value переменной, но не совсем поняла, как он работает.

``
((JavascriptExecutor) driver).executeScript("document.getElementByXpath(\"//*[@id='root']/div/div[2]/div[3]/button[2]\").setAttribute('value', duration)");
```
Прошу прощения (слезно!!!) за позднюю сдачу, но к концу спринта мозг знатно устал обрабатывать информацию.

Тест с выпадающим списком, кстати, вроде работает!

Хорошего дня, надеюсь на скорый фидбэк!
Спасибо!

Commit:
Превая версия финального проекта 4 спринта.

pavel-ulan commented on src/test/java/listOfAnswers.java:
```diff
@@ -0,0 +1,43 @@
+        this.driver = driver;
+    }
+
+    private By firstQuestion = By.xpath("//*[@id=\"root\"]/div/div/div[5]/div[2]/div/div[1]");
```
**надо исправить**
очень очень некрасивый локатор, состоящий из только из пути =/

Старайся не использовать локаторы состоящие только из пути (не важно относительный/абсолютный). Как правило всегда можно подобрать другой, более читаемый, локатор. Такие локаторы плохо читаются, по ним непонятно, куда они указывают и если что-то меняется в верстке, то они ломаются и могут или никуда не вести или, что еще хуже, возвращать совершенно другой элемент. 

Пример: кто-то сменил верстку сайта (незначительно. пару новых div добавил). И вот твой путь внезапно начал вести к совершенно другому полю, а оно тоже div. И у тебя тесты начинают давать ложные результаты. Дальше события могут развиваться по разному, но при худшем раскладе в прод может улететь версия с непротещенной функциональностью. А если она еще и сломалась...

Ну и плюс по такому локатору фиг поймешь к чему он ведет (кроме типа поля).

Так что локаторы я попрошу тебя переделать.

pavel-ulan commented on src/test/java/orderDetails.java:
```diff
@@ -0,0 +1,44 @@
+    public orderDetails(WebDriver driver) {
+        this.driver = driver;
+    }
+    private By dateOfOrder = By.xpath("//*[@id='root']/div/div[2]/div[2]/div[1]/div[1]/div/input");
```
ну и эти тоже переделать

pavel-ulan commented on src/test/java/Praktikum.java:
```diff
@@ -0,0 +1,94 @@
+            objOrderDetails.waitForLoadOrderButton();
+            objOrderDetails.setOrderDetails(date,duration);
+            objOrderDetails.clickColorOfItemIsBlack();
+            objOrderDetails.clickOrderButton();
```
**надо исправить**
а где собственно проверка? не забывай, тест это не сценарий. Тест должен что-то проверять.

pavel-ulan reviewed (changes requested):
Смотри, помимо того, что я отметил комментами, у тебя не хватает тестов. Вопросы должны проверяться все (а не 1 и последний). Ну и заказ должен проверяться для верхней и средней кнопки (а у тебя все 3 параметра только для верхней кнопки заказать работают.

Касательно твоего вопроса - driver.findElement(undergroundStation).sendKeys(station) работает для поля input.  А выбор станции это дропдаун. Можно выбрать и через jsscript, а можно и через webDriver. Погугли как работать с дропдаун полями, думаю найдешь быстро (там ничего сложного нет)

Commit:
Исправленны замечания к пулл-реквесту.
Добавлены тестовые параметры для теста проверки раздела "Вопросы о важном".
Доработаны тесты заказа самоката.

Commit:
.gitignore updated

Commit:
Добавлен .gitignore
Скрыты ненужные файлы

Commit:
FINALLY!
Исправлены все тесты, добавлен .gitignore, скрыты все лищние файлы и исправлен README.md

pavel-ulan commented on src/test/java/Praktikum.java:
```diff
@@ -0,0 +1,143 @@
+                objListOfAnswers.clickToGetAnswer(question);
+                String actual = objListOfAnswers.getTextAnswer(answer);
+                assertEquals(answer, actual);
+                driver.quit();
```
**надо исправить**
у тебя есть в after методе закрытие драйвера, зачем он тебе тут? 
закрывать драйвер в тесте - неправильно. если тест падает - драйвер не закроется, а это путь к утечке данных и перегрузке CI системы.
убери его отсюда. закрытие в after блоке - верное и достаточное решение

pavel-ulan replied:
ну и из-за того, что ты его тут закрываешь у тебя и была ошибка в after блоке (она же у тебя была?). т.к. ты пыталась в after закрыть уже несуществующий драйвер

pavel-ulan commented on src/test/java/orderForm.java:
```diff
@@ -0,0 +1,58 @@
+    public orderForm(WebDriver driver) {
+        this.driver = driver;
+    }
+    private final By nameField = By.xpath("//*[@id='root']/div/div[2]/div[2]/div[1]/input");
```
**надо исправить**
а тут исправить локаторы?

pavel-ulan commented on src/test/java/orderForm.java:
```diff
@@ -0,0 +1,58 @@
+    private final By nameField = By.xpath("//*[@id='root']/div/div[2]/div[2]/div[1]/input");
+    private final By surnameField = By.xpath("//*[@id='root']/div/div[2]/div[2]/div[2]/input");
+    private final By addressField = By.xpath("//*[@id='root']/div/div[2]/div[2]/div[3]/input");
+    //private String xpathUnderground = ".//div[@class='select-search__select']//*text()=\"
```
**надо исправить**
не оставляй ненужных комментариев или закомменченный код в финальной работе

pavel-ulan reviewed (changes requested):
Молодцом, теперь работа действительно выглядит как надо =)
Но (куда ж без него) есть несколько упущений в оформлении. все в комментах. посмотри, поправь и давай еще раз глянем =)

Commit:
Тесты разделены по пакетам, добавлен enum для кросс-браузерного тестирования.
Исправлены ошибки с использованием аннотации @After для закрытия браузера

pavel-ulan commented on src/test/java/ListOfAnswersTest.java:
```diff
@@ -0,0 +1,70 @@
+                {browserList.CHROME, "Можно ли продлить заказ или вернуть самокат раньше?", "Пока что нет! Но если что-то срочное — всегда можно позвонить в поддержку по красивому номеру 1010."},
+                {browserList.CHROME, "Вы привозите зарядку вместе с самокатом?", "Самокат приезжает к вам с полной зарядкой. Этого хватает на восемь суток — даже если будете кататься без передышек и во сне. Зарядка не понадобится."},
+                {browserList.CHROME, "Можно ли отменить заказ?", "Да, пока самокат не привезли. Штрафа не будет, объяснительной записки тоже не попросим. Все же свои."},
+                {browserList.CHROME, "Я жизу за МКАДом, привезёте?", "Да, обязательно. Всем самокатов! И Москве, и Московской области."}
```
**можно лучше**
ок, как ошибку уже считать не буду, НО... Браузер - это не параметр теста. браузер это параметр запуска. т.е. в каком браузере будет все запускаться должно определять до запуска тестов (а у тебя это один из параметров.).
тут проще всего было бы в Before сделать что-то типа 
```
getInstance() {
            if (browser == "firefox") {
                WebDriverManager.firefoxdriver().setup();
                INSTANCE = new FirefoxDriver();
            } else {
                WebDriverManager.chromedriver().setup();
                INSTANCE = new ChromeDriver();
            }
        }
        return INSTANCE;
    }
```
и просто передавать browser (хоть в виде константы, хоть в виде параметра из property файла).

pavel-ulan replied:
У тебя даже реализация по сути есть в BrowserSwitch (кстати, очень даже хорошая молодцом). Просто надо было вынести ее в Before метод или BaseTest, а не пытаться ее запихнуть в параметры теста.

pavel-ulan reviewed (approved):
Молодцом, все еще есть к чему стремиться, но уже значительно лучше. Немного перемудрила с браузером конечно.
